Userland/Libraries/LibGfx/ImageFormats/PNGWriter.cpp

```diff
@@ -326,4 +391,99 @@ ErrorOr<ByteBuffer> PNGWriter::encode(Gfx::Bitmap const& bitmap, Options options
     return stream.read_until_eof();
 }
 
+class PNGAnimationWriter : public AnimationWriter {
+public:
+    PNGAnimationWriter(SeekableStream& stream, IntSize dimensions, int loop_count, PNGWriter::Options const& options)
+        : m_writer(stream)
+        , m_stream(stream)
+        , m_dimensions(dimensions)
+        , m_loop_count(loop_count)
+        , m_options(options)
+    {
+    }
+
+    virtual ErrorOr<void> add_frame(Bitmap&, int, IntPoint, BlendMode) override;
+
+private:
+    PNGWriter m_writer;
+    SeekableStream& m_stream;
+
+    IntSize const m_dimensions;
+    int const m_loop_count { 0 };
+
+    bool m_is_first_frame { true };
+
+    u32 m_sequence_number { 0 };
+    u32 m_number_of_frames { 0 };
+    size_t m_acTL_offset { 0 };
+    PNGWriter::Options const m_options;
+};
+
+ErrorOr<void> PNGAnimationWriter::add_frame(Bitmap& bitmap, int duration_ms, IntPoint at, BlendMode)
+{
+    ++m_number_of_frames;
+    bool const is_first_frame = m_number_of_frames == 1;
+
+    if (is_first_frame) {
+        // "The fcTL chunk corresponding to the default image, if it exists, has these restrictions:
+        //  * The x_offset and y_offset fields must be 0.
+        //  * The width and height fields must equal the corresponding fields from the IHDR chunk."
+        // FIXME: If this ends up happening in practice, we should composite `bitmap` to a temporary bitmap and store that as first frame.
+        if (at != IntPoint {})
+            return Error::from_string_literal("First APNG frame must have x_offset and y_offset set to 0");
+        if (bitmap.size() != m_dimensions)
+            return Error::from_string_literal("First APNG frame must have the same dimensions as the APNG itself");
+
+        // All frames in an APNG use the same IHDR chunk, which means they all have the same color type.
+        // To decide if we should write RGB or RGBA, we'd really have to check all frames, but that needs a
+        // lot of memory and makes streaming impossible.
+        // Instead, we always include an alpha channel. In practice, inter-frame compression means that
+        // even for animations without transparency, all but the first frame will have transparent pixels.
+        // The APNG format doesn't give us super great options here.
```

Comment:
In which scenario, following frames have a different color type than the first one?
I think that this is ok to assume that they will have the same type. IMO, applications like PixelPaint should just not do stuff like that. And if we pass multiple single frame images with different color types as an input of `animation`, simply showing a warning seems good enough to me.

It seems to me that we are bloating every single generated image just in case the API is used in a very weird way, so I'm not sure if it's really worth.

Reply:
Imagine an animation where some frames have transparent pixels and some don't, or some frames have more than 256 colors and some don't. If we could store color types per-frame, we could store the frames that don't have transparent pixels as just RBG instead of RGBA, or the frames with <= 256 colors as indexed. That would could save quite a bit of storage. But since we can't do per-frame image formats, we have to store full RGBA even for frames that don't need it.

(Along the same lines, when an image only has a single, fully transparent color, like a decoded gif always has, we could store image data for as RGB + `tRNS` chunk instead of storing full RGBA – but again, we can't really do this for animations.)